scipy/sparse/linalg/eigen/arpack/arpack.py

```diff
@@ -1231,11 +1231,31 @@ def eigs(A, k=6, M=None, sigma=None, which='LM', v0=None,
             import warnings
             warnings.warn('M does not have the same type precision as A. '
                           'This may adversely affect ARPACK convergence')
+
     n = A.shape[0]
 
-    if k <= 0 or k >= n:
-        raise ValueError("k=%d must be between 1 and ndim(A)-1=%d"
-                         % (k, n - 1))
+    if k <= 0:
+        raise ValueError("k=%d must be greater than 0." % (k, n - 1))
+
+    if k >= n - 1:
+        if not isdense(A):
+            raise TypeError("Using scipy.linalg.eig instead as k >= N - 1. "
+                            "Hence, 'A' cannot be sparse. "
+                            "Hint: Use A.toarray() before passing to the "
+                            "function.")
+
+        import warnings
+        warnings.warn("k >= N - 1 for N * N square matrix. "
+                      "Using scipy.linalg.eig instead.", UserWarning)
+
+        if isinstance(A, LinearOperator):
+            raise TypeError("Using scipy.linalg.eig instead as k >= N - 1. "
+                            "Hence, 'A' cannot be a LinearOperator")
+        if isinstance(M, LinearOperator):
+            raise TypeError("Using scipy.linalg.eig instead as k >= N - 1. "
```

Comment:
These `TypeErrors` say that they're using `scipy.linalg.eig`, but they're not. Maybe something like:

Cannot use scipy.linalg.eigs for sparse A with k >= N - 1. Use scipy.linalg.eig(A.toarray()) or reduce k.

Cannot use scipy.linalg.eigs for LinearOperator A with k >= N - 1.